doc/neps/nep-0049.rst

```diff
@@ -0,0 +1,369 @@
+enable exploration of this rich area of possible optimizations. The intention
+is to create a flexible enough interface without burdening normative users.
+
+_`issue 5312`: https://github.com/numpy/numpy/issues/5312 
+_`from 2017`: https://github.com/numpy/numpy/issues/5312#issuecomment-315234656
+_`in 2005`: https://numpy-discussion.scipy.narkive.com/MvmMkJcK/numpy-arrays-data-allocation-and-simd-alignement
+_`here`: http://numpy-discussion.10968.n7.nabble.com/Aligned-configurable-memory-allocation-td39712.html
+_`and here`: http://numpy-discussion.10968.n7.nabble.com/Numpy-s-policy-for-releasing-memory-td1533.html
+_`issue 14177`: https://github.com/numpy/numpy/issues/14177
+_`filprofiler`: https://github.com/pythonspeed/filprofiler/blob/master/design/allocator-overrides.md
+_`electric fence`: https://github.com/boundarydevices/efence
+_`memory interfaces`: https://docs.python.org/3/c-api/memory.html#customize-memory-allocators
+_`BPO 18835`: https://bugs.python.org/issue18835
+_`PEP 445`: https://www.python.org/dev/peps/pep-0445/
+
+Usage and Impact
+----------------
+
+The new functions can only be accessed via the NumPy C-API. An example is
+included later in the NEP. The added ``struct`` will increase the size of the
+``ndarray`` object. It is a necessary price to be paid for this approach. We
+can be reasonably sure that the change in size will have a minimal impact on
+end-user code because NumPy version 1.20 already changed the object size.
+
+The implementation preserves the use of ``PyTraceMalloc_Track`` to track
+allocations already present in NumPy.
+
+Backward compatibility
+----------------------
+
+The design will not break backward compatibility. Projects that were assigning
+to the ``ndarray->data`` pointer were already breaking the current memory
+management strategy (backed by ``npy_alloc_cache``) and should restore
+``ndarray->data`` before calling ``Py_DECREF``. As mentioned above, the change
+in size should not impact end-users.
+
+Detailed description
+--------------------
+
+High level design
+=================
+
+Users who wish to change the NumPy data memory management routines will use
+:c:func:`PyDataMem_SetHandler`, which uses a :c:type:`PyDataMem_Handler`
+structure to hold pointers to functions used to manage the data memory. The
+calls are wrapped by internal routines to call :c:func:`PyTraceMalloc_Track`,
+:c:func:`PyTraceMalloc_Untrack`, and will use the
+:c:func:`PyDataMem_EventHookFunc` mechanism  already present in NumPy for
+auditing purposes. 
+
+Since a call to ``PyDataMem_SetHandler`` will change the default functions, but
+that function may be called during the lifetime of an ``ndarray`` object, each
+``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
+time of its instantiation, and these will be used to reallocate or free the
+data memory of the instance. Internally NumPy may use ``memcpy` or ``memset``
+on the data ``ptr``.
```

Comment:
nitpick
```suggestion
Since a call to :c:func:`PyDataMem_SetHandler` will change the default functions, but
that function may be called during the lifetime of an ``ndarray`` object, each
``ndarray`` will carry with it the :c:type:`PyDataMem_Handler` struct used at the
time of its instantiation, and these will be used to reallocate or free the
data memory of the instance. Internally NumPy may use ``memcpy` or ``memset``
on the data ``ptr``.
```

Reply:
Another question: is ``ptr`` a member of some struct?

Reply:
I find the multiple use of links distracting. The first mention is a link, the others are not.

> is `ptr` a member of some struct?

No, it is the return value from `alloc`. I will reformat to make it more clear.